Fix handles being split by `asset_server_managed`.

andriyDev opened the pull request:
# Objective

- Fixes #20651.
- Alternative to #20874.
- Also a step towards #21878.

## Solution

- When allocating a new handle (whether for an existing asset or a new asset), send a "new" message to the same channel as "drop" events.
    - This essentially means we're "leaning in" to the fact that we can have multiple live handles to the same asset.
- `Assets::track_assets` now increments when a "new" message is received, and decrements when a "drop" message is received. The asset entry is dropped if the counter is at 0 after processing all messages.
- Remove the `asset_server_managed` flag, since we can just lookup the index to see if it's present.
- Remove the `handle_drops_to_skip` counter since now we count all new handles, not just those in `get_strong_handle`.

## Testing

- Unbroke our reproduction test.
- Added a new test to verify that "reacquiring" an asset doesn't drop the asset.

kristoff3r reviewed (approved):
LGTM, the logic in `track_assets` felt a bit complicated but I couldn't find a way to simplify it.

andriyDev commented:
@janis-bhm That was a good catch! I didn't know these methods existed. Now when you call `AssetServer::get_id_handle`, it will allocate a new handle if possible. We now ignore the weak_count entirely and just say if the entry still exists, you can get a handle to it. I also entirely removed the `is_path_alive`, inlined it in the only place its used, and simplified it to assume that if the entry exists, the path still exists.

@kristoff3r I've added some not entirely trivial changes, so I'd like another review just in case.

andriyDev commented on crates/bevy_asset/src/handle.rs:
```diff
@@ -19,60 +18,63 @@ use uuid::Uuid;
 #[derive(Clone)]
 pub struct AssetHandleProvider {
```
I am of the opposite opinion - handles should store almost no state IMO. That's because if in the future we want to support no_std, it may be the case that we allow users to *entirely disable* the AssetServer and allow users to "bring their own asset server", in which case, relying too much on state in the Handle won't work (unless we allow adding arbitrary state to a handle which is a little gross).

Also: the prior attempt at this was https://github.com/bevyengine/bevy/pull/20874 which just used a Mutex<HashMap> to avoid creating duplicate handles. Alice was not excited by the extra locking, though I don't think that's a big deal since the AssetServer already locks anyway.

cart replied:
In an "assets as entities" world I do see a path where Handles are "just" handles to a specific Entity, which is decoupled from the asset system entirely (I've built that before actually). But even in that world, a single-shared canonical handle is valuable, as it allows us to easily count references (useful for debugging), and it allows us to treat the dropping of the handle as the "end" of the Entity. It also allows us to maximize the lifetime of Weak handles: if there is only ever one handle lifetime, then we can always successfully upgrade the Weak handle, provided there is still one active reference somewhere. In multi-handle, Weak::upgrade failing doesn't necessarily mean that the asset (or entity) is dropped. Plus multi-lifetime just requires more infrastructure / complexity / overhead and introduces more surface area for weirdness.

andriyDev replied:
@cart That is indeed a simple policy but loses out on a nice benefit of the current approach: if you drop the last handle to an asset, but then call `load` in the same frame, you can "revive" that asset (it won't trigger another load and will just keep the existing asset). Maybe that's not that big of a win? For some reason I was really worried about losing this, but I don't remember why...

If that's a benefit we don't care about, then going to a simpler model seems fine. But I think it's worth making sure we don't care about this (or that there aren't obvious pitfalls).

kristoff3r reviewed (approved):
Small nit for the inner loop, but other than that the PR still seems good

janis-bhm commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -312,61 +312,63 @@ impl AssetInfos {
-        let info = self.infos.get(&index)?;
-        let strong_handle = info.weak_handle.upgrade()?;
-        Some(UntypedHandle::Strong(strong_handle))
+    pub(crate) fn get_index_handle(&mut self, index: ErasedAssetIndex) -> Option<UntypedHandle> {
```
Maybe this could get a comment explaining why it's `&mut` now? `get_index_handle` sounds like some immutable operation, but since the asset behind the index may no longer have a handle, this could actually create a new handle.
Publicly facing, this I think only results in `AssetServer::get_handle` and friends making an asset live-again if the method is called on a "dangling" asset, which I think is actually the desired behaviour.

janis-bhm reviewed (approved):
LGTM.

I've added a few comments mainly just documenting my thought process and a few nits, I guess.
